### ISSUE TITLE
Magit 2.1.0 will be released on the 1th of July

### ISSUE THREAD

**Issue by tarsius** [magit/magit#1645]
## Magit 2.1.0 will be released on the 1th of July

----

**[Related issues](https://github.com/magit/magit/milestones/2.1.0)** | **[Update instructions](http://magit.vc/manual/magit/Installation.html)** | **[MAGIT 2.1.0 RELEASED](https://github.com/magit/magit/issues/1941)**

----

## Release steps

**Bold** items are important check-points.
_Italic_ items require that someone lends me a hand.

* [x] **announce release date**
* [x] inform maintainers of some
  * [x] extensions #1921
  * [x] themes #1922 
  * [x] starter-kits #1923
* [x] prepare pull requests
  * [x] melpa milkypostman/melpa#2892
  * [x] solarized bbatsov/solarized-emacs#189
  * [x] zenburn bbatsov/zenburn-emacs#200
  * [x] magit-gh-pulls sigma/magit-gh-pulls#31
* [x] test elpa packaging using fake release and local server
  * [x] melpa
  * [x] marmalade
  * [x] update without following instructions (went better than expected)
* [x] copy fixes from manual to respective blurbs elsewhere
* [x] update information and links on webpage and wiki
* [x] strip temporary information from README.md
* [x] regenerate Authors.md
* [x] finish release notes
* [x] **merge `next` branch into `master`**
  * [x] git-modes
  * [x] magit
* [x] rename `next` branch to `master`
  * [x] magit-stgit
  * [x] magit-svn
  * [x] magit-topgit
* [x] tag
  * [x] git-modes 1.2.0
  * [x] magit 2.1.0
  * [x] magit-stgit
  * [x] magit-svn
  * [x] magit-topgit
* [x] **push branch and tag**
* [x] _get melpa recipes merged_
* [x] _get themes merged_
  * [x] _solarized_ bbatsov/solarized-emacs#189
  * [x] _zenburn_ bbatsov/zenburn-emacs#200
* [ ] _get `next` of extensions merged_
  * [x] magit-annex https://github.com/magit/magit-annex/commit/cf1f2cfe9ed20564c7b5ecdeb8b4cc2d0f89936b
  * [ ] ...
* [x] _get milkypostman/melpa#2641 closed_
* [ ] upload to marmalade
  * [ ] magit
    * [x] with-editor
    * [ ] _get git-commit uploaded_ nicferrier/elmarmalade#110
    * [x] magit-popup
    * [x] magit
  * [ ] git-modes
    * [x] gitattributes-mode
    * [ ] _get gitconfig-mode uploaded_ magit/git-modes#126
    * [x] gitignore-mode
  * [ ] _get third-party extensions uploaded_
  * [ ] _get obsolete git-commit-mode and git-rebase-mode removed_ nicferrier/elmarmalade#111
* [x] wait for changes to propagate
* [x] **inform elpa users that it is safe to update**
  * [x] melpa
  * [ ] _marmalade_ (either Nic gets his act together or this will never happen)

## Post release

* [ ] **help users who did not uninstall the previous installation before updating**
* [ ] update screenshots and cheat sheet
* [ ] remind maintainers of extensions and themes to update
* [ ] _discuss deprecation of packages_
* [ ] _ask Nic to transfer marmalade package permissions to me_ (once more: magit/git-modes#126, nicferrier/elmarmalade#101)
* [ ] _ask Nic to remove git-commit-mode and git-rebase-mode_
* [ ] rename `later (see #1645)` milestone to just `later`
* [ ] adjust answers on [Stackexchange](http://emacs.stackexchange.com/questions/tagged/magit)

**Comment by phil-s** [comment]
Having just switched to `next`, I immediately found myself slightly annoyed by the section cycle behaviour.

FWIW, I also couldn't see how to configure it! (I was looking for applicable variables, rather than functions to rebind the key to), and I suspect others will be in the same boat.

Cycling is great for some things, but less so (to me) for others. I'd like to propose an added level of control over the behaviour. Sections have a type, and it seems to me that being able to configure the behaviour on a per-type basis would be ideal.

At minimum, the ability to select 'toggle' or 'cycle' would be useful. Potentially the ability to specify the cycle sequence would be good as well (such that you could go from fully-collapsed -> fully-expanded -> hidden-children; or start by default at hidden-children, and then -> fully-expanded -> fully-collapsed), but I'm not sure whether that's something I actually want, or just something that sounds good in terms of providing flexibility.

**Comment by tarsius** [comment]
While things have been a bit slow and no items have been ticked of there was some progress. A lot of small issues have been fixed. Almost all doc-strings are complete now -- I can only think of one command which definitely needs work on the doc-string. Of course they can, and probably will, be further improved, but "Add an accurate doc-string to every command and every option" can almost be ticked of now.

I have also implemented a few new features. In most cases these were obvious extensions to existing features, but there also were a few others I have to admit. In most cases *those* features were things I personally need, e.g. the new submodule commit sections. I use submodules to install Emacs packages, and these new sections make a huge difference. I also implemented some newly proposed features which I shouldn't have done, and will try harder to refuse to do in the future.

Out of the existing outstanding feature request one I might implement soon (#1566), one I hope someone else implements (#1638), one I am tempted to implement but really shouldn't (#1485), while the remaining four most definitely won't make it into the release (#1615, #1586, #1425, #1418).

**Comment by tarsius** [comment]
> Having just switched to next, I immediately found myself slightly annoyed by the section cycle behaviour.

I just tried that and now that I am used to cycling I immediately found myself slightly annoyed by the section **toggling** behaviour :-)

Before I tried that I considered using `SPC` for toggling, because I think consistency with Org-mode would be a good thing and that uses `TAB` for cycling. But `SPC` cannot be used for toggling because in some places it is used for scrolling and I don't think there is a good alternative key that could be used.

> FWIW, I also couldn't see how to configure it! (I was looking for applicable variables, rather than functions to rebind the key to), and I suspect others will be in the same boat.

Maybe, but I personally hate to add unnecessary options -- they always tend to grow a live of their own. Also as someone who heavily customizes key bindings, I find that such key-binding options make it *harder* to do so. It's not absolutely out of the question, but I have very strong reservations.

Or is

```lisp
(define-key magit-mode-map [C-tab] 'magit-section-toggle)
(define-key magit-mode-map [tab] 'magit-section-toggle)
```

really that much worse than

```lisp
(setq magit-section-tab-does 'toggle)
```

?

> Cycling is great for some things, but less so (to me) for others. I'd like to propose an added level of control over the behaviour. Sections have a type, and it seems to me that being able to configure the behaviour on a per-type basis would be ideal.

First of all, that's already possible. Sections also have a dedicated keymap `magit-TYPE-section-map` which can be used to accomplish this. Of course that again requires the use of `define-key` but I think that is an argument *against* adding an option to control whether cycle or toggle is used. Because if you want to (a) use an option (b) allow setting behaviour per section type, you suddenly need *many* options.

Also I think we should stick to one behaviour by default, anything else would be very confusing. So I think we should stick to the current, and very standard, configuration mechanism (bindings in keymaps).

There are also not many sections where differentiating between toggling and cycling even makes a difference. Mostly it's just the diffs really, and here, now that I have used it for a while, I have come to favor cycling.

Another place I can think of is the refs buffer (`y`), here you have sections for remotes, which contain the respective branches, which can be expanded to show the commits not merged into the current branch. Showing the commits for all branches under a remote at once would be rather expensive, because we show them as "cherries", i.e. we check whether the commit is "completely new" or exists on the current branch under a different rev.

Luckily (well, not exactly luck, I implemented it that way) cycle does the right thing in that case. There are more than two section states expanded and collapsed. There's also the trivial case of a section that has not body, which isn't relevant here. And "might have a body, but we won't know until we try expanding it". This state exists to avoid expensive work which only becomes necessary if the user actually dives in that deep. Cycle only attempts to expand such a section, if it is invoked directly on that section, if it is invoked on a parent of such sections, then it behaves like toggle.

What the default bindings should be is a different question. You should honestly ask yourself why you are suggesting the addition of an option. I don't want to imply any "bad" intentions, it's just that in the past users, once they realized that I am not going make their preference the default, very often switch to "well then at least make it customizable". So do you really think it should be customizable regardless of what the default is, or do you just a quick toggle (he!) to make it the way you think it ought to be in the first place. Adding an option isn't "at least" to me -- on the contrary it very often is what I consider to be the worst possible outcome.

Likewise I have asked myself why have such strong reservations towards making everything customizable. Of course in some cases I just feel that the proposed alternative behaviour just doesn't make much sense, but often it's also because of the added complexity and the fear of it turning into a maintenance nightmare. In each case the complexity would only increase only a little, but if I did it everytime someone asked me to, then that would be a whole other story. The addition of an option can cause a lot of work down the road once someone comes along suggesting that it shouldn't be a boolean because that other variant would also make sense ... So when it is not necessary to add an option, and here it is not, `define-key` works just fine, then I don't add an option.

> At minimum, the ability to select 'toggle' or 'cycle' would be useful. Potentially the ability to specify the cycle sequence would be good as well (such that you could go from fully-collapsed -> fully-expanded -> hidden-children; or start by default at hidden-children, and then -> fully-expanded -> fully-collapsed), but I'm not sure whether that's something I actually want, or just something that sounds good in terms of providing flexibility.

Probably the latter.

> I think the simpler (toggle-or-cycle) version would be a good thing to include for release.

Let's stick to `define-key`.

**Comment by tarsius** [comment]
Just to make it clear: I still haven't decided whether toggle or cycle get's to keep `TAB`, just that there won't be any options controlling this.

**Comment by tarsius** [comment]
I am going on vacation soon. I might, or might not, find some time to comment on issues again before then.

When I get back I will probably be all exited to hack on Magit, including documentation even. I would hate to see that energy be drained by conversations along the line "sorry we are in feature freeze", "but this is really important and not much work", so I would really appreciate it if you guys could inform people who open new feature requests while I am gone about the feature freeze and that it's actually kinda important that I get more dogmatic about it. Thanks.

Also if #1638 were a solved issue when I get back, that would be great :-)

**Comment by phil-s** [comment]
Thanks for the detailed response, tarsius. Having read all that, I think it's actually just a question of documentation. With per-section-type keymaps, the flexibility is already there, so I agree that there's no necessity for additional configuration.

I'm not really sure what the default binding should be, as clearly there are arguments for both options. If cycling remains as the default, then I think the release notes will need to highlight the change and explain how to configure toggling (on a per-type basis).

Have a great holiday!

**Comment by tarsius** [comment]
Thanks to everyone who helped users while I was gone. There are still quite a few new issues and it will take some time to get up to speed again.

**Comment by tarsius** [comment]
For the first time the only remaining issues in the `2.1.0` milestone are this one and the "there be documentation" issue. Jupiiiii.

**Comment by Silex** [comment]
\o/

Ship it!!! :laughing:

**Comment by syl20bnr** [comment]
That's cool !!

**Comment by bbatsov** [comment]
Nice! Looking forward to 2.1! :-)

**Comment by bradleywright** [comment]
:+1: :+1: :+1:

**Comment by jwiegley** [comment]
I'd enjoy having cycling as an optional behavior, rather than the default.  When I want to see hunk titles, I press '2', which is rarely.

**Comment by tarsius** [comment]
@jwiegley Errr, you might want to mention your argument :-)

> The Org-like functionality is ok, except that in Org, you *care* about the heading sometimes more than the content, whereas in Git, I'd think you care about content more than the headings.

**Comment by jwiegley** [comment]
That's all: my argument is that the first thing you want to see after pressing TAB in a diff buffer is the diffs, rather than a summary of the diff structure.  Whereas when I open an Org file, I often want to see the structure first so that I can decide which aspect of the content to view.

**Comment by tarsius** [comment]
Yes, but except for me nobody saw that before I quoted it here :-)

**Comment by jwiegley** [comment]
Oh, I didn't even realize you had quoted something from another source!  Thanks. :)

**Comment by tarsius** [comment]
/rant

When people suggested that texinfo be replaced with something else, I did not understand what all the whining was about. Now I do.

For example it's not possible to create a crossreference, without texinfo injecting the string "see" in front of it. So I either have to reword the sentence so that the "see" makes sense (but then I sometimes end up with see X, see Y, and see Z, when I would have preferred see X, Y, and Z). Or put all links in footnotes. Sight. I understand why it is done like this (printed manual) but it is still stupid (and the output is not only not what I want in info, in the pdf it's also not what I want).

Of course I don't use texinfo directly but author the manual in org, but that does not really make things easier. E.g. this does not allow me to work around the crazy things texinfo does to links, instead I have to anticipate for that stupidity while I am in org - and that's a bit sad because org excels at links.

Oh and `ox-texinfo` is very incomplete, so I had to write [`ox-texinfo+`](https://github.com/tarsius/ox-texinfo-plus).

Compromises had to be made, but finally I am starting to get somewhere.

**Comment by tarsius** [comment]
I have flipped the bindings for `magit-section-toggle` and `magit-section-cycle`, mostly because if I did not, then I would have to justify why I think that the latter should get the better binding (I still do) over and over again.

**Comment by npostavs** [comment]
> For example it's not possible to create a crossreference, without texinfo injecting the string "see" in front of it.

That seems so ridiculous, I looked in the texinfo manual to find a way around it, [`@ref`][ref] seemed promising:

> `@ref` is nearly the same as `@xref` except that it does not generate a ‘See’ in the printed output, just the reference itself. This makes it useful as the last part of a sentence. 

but

>     For more information, @pxref{This}, and @ref{That}.
> 
> produces in Info:
> 
>     For more information, *note This::, and *note That::.
> 
> and in printed output:
> 
>     For more information, see Section 1.1 [This], page 1, and Section 1.2 [That], page 2.

`*note That::` renders as `see That` in the info viewer. So you can omit the "see" in the printed output, but not in the info. WTF?


[ref]: https://www.gnu.org/software/texinfo/manual/texinfo/html_node/_0040ref.html

**Comment by tarsius** [comment]
The should probably add `@lref`:

> `@lref` is nearly the same as `@xref` and `@ref` except that it actually does what you want it to do. `l` stands for "literally what I told you to do". It takes three arguments; the link, the text displayed as a link in info, the text displayed in printed manuals. The third argument can be omitted in which case the same text is used for info and dead trees. Since dead tree documents don't support links, the text you provided it followed by something like " (Section N, page M)".

And that doesn't even take into account that most pdfs are not actually printed out and that "electronic pdfs" actually *do* support links.

**Comment by tarsius** [comment]
> `*note That::` renders as `see That` in the info viewer. 

Actually, since this is Emacs, we might be able to fix that by redefining `Info-fontify-node` to allow packages to opt-in to not ever have "see " inserted before links in their manuals. But having to deal with grap like that of course doesn't help getting the manual done.

**Comment by jwiegley** [comment]
@tarsius And I will thank you forever (for flipping the bindings).

**Comment by tarsius** [comment]
Magit now has a new no-nonsense web presence at http://magit.vc.
The draft for the `2.1.0` is now available from there.

**Comment by tarsius** [comment]
I released `1.4.0` based on the `master` branch.

**Comment by tarsius** [comment]
The `next` branch is now the HEAD/default branch for Magit and the other repositories which have a `next` branch of their own. @kyleam: that includes magit-annex.

The idea behind this move that it will encourage more experienced users to use the `next` branch. Moving everyone to `next` would not be a good idea because the documentation still isn't ready, but some additional beta testers should actually help finding bugs in features the current `next` users don't use often.

Lately I was busy getting the new webpage and `1.4.0` ready, so a few new issues have piled up. I should get to that soon, but will probably first have to deal with problems that users updating to `1.4.0` might have. Also I might write some paragraphs in the manual first...

**Comment by phil-s** [comment]
Quick heads-up that on 'next' magit doesn't recognise repositories if you're running git version 1.7.9.5 (I don't know which version marks the tipping point; this is just the version I had on one of my machines).

I think you'll want to have some kind of user notification regarding the minimum supported version of git for Magit 2.x, and this should be clearly noted as a pre-requisite in the installation instructions.

**Comment by tarsius** [comment]
Magit started complaining about the local Git being to old a few days ago. Does anyone know of a "when first connecting to a host using tramp, but only once for each host" hook that I could use to do the same for the Git versions on remote machines?

**Comment by tarsius** [comment]
> this should be clearly noted as a pre-requisite in the installation instructions.

Well, it's not *in* the installation instructions, but one centimeter above :-)

**Comment by vermiculus** [comment]
> Does anyone know of a "when first connecting to a host using tramp, but only once for each host"-hook that I could use to do the same for the Git versions on remote machines?

Nothing shows up with `helm-describe-function ^tramp hook`, but I imagine you could `defadvice` Tramp's connection function. That would be a good Emacs.SE question.

**Comment by phil-s** [comment]
> Well, it's not in the installation instructions, but one centimeter above :-)

Ooh. So it is. I managed to miss it, though -- if all you're looking for is installation instructions, it's pretty easy to start reading at the big "Installation" heading.

I'd also mention it in here:
https://github.com/magit/magit/wiki/Installation#installing-from-the-next-branch

I see your nice detailed in-Emacs warning code now. It was in the version I was running, but you're using after-init-hook which won't trigger it if Magit is (auto)loaded subsequent to that.

**Comment by tarsius** [comment]
Yeah, the instructions were incomplete, so I am switching to just providing it in one place for the time being, on the wiki.

**Comment by yashi** [comment]
a small hiccup. magit next (or `version-to-list`) complains if you are using development version of git:
`version-to-list: Invalid version syntax: '2.4.0.rc3.3.g6eb1401'`

this happens if you `(require 'magit)` in your init file.  I'll open an issue if needed.

**Comment by tarsius** [comment]
> a small hiccup.

Should work now.

**Comment by tarsius** [comment]
On of these days ... I am hammering out fixes, but they keep coming at the same rate.

**Comment by tarsius** [comment]
I'll be offline for most of the next two weeks. I might show up once or twice though.

**Comment by tarsius** [comment]
So the `next` branch is ready for release. Now I just have to resist the temptation to replace `magit-popup.el` with yet another implementation, and instead finish the documentation.

That's actually related, most parts of the documentation that currently consist of just a list of relevant symbols are like that because they have to do with popups. Those are pretty important parts of the documentation, but I avoided working on them because I am uncomfortable with the term "popup". The old dominant term when talking about this feature was "key mode", which made even less sense because in Emacs pretty much everything is a "key mode". But popup also isn't very nice because it focuses on the visual part instead of what the feature is about, i.e. infix arguments.

I have talked about a successor to `magit-popup.el` for a while now, the working title was `idefix.el`. I have implemented a very incomplete prototype but finishing it now would be to much work - it would delay the release yet again. But I intend to focus on this soon after the release.

The new implementation will be named `transient.el` and be a synthesis of "prefix argument" and "prefix commands".  It will provide three macros used for defining commands `define-transient-command`, `define-infix-argument`, and `define-suffix-command`. The implementation will center around the use of a transient keymap. Infix arguments and suffix commands will likely have an associated eieio object. It will be possible to define any number of argument classes, not just "options" and "switches". And the default state of the arguments will take the appropriate (Git (in Magit's case)) variables into account. Stuff like that. I don't want to go into details now, that's not really the topic right now. Anyway it will be awesome :-)

`magit-popup` has been available from Melpa as a separate package for a while now. But there never was an actual release. I would like to rename it before `magit-2.1.0` and `package-formerly-known-as-magit-popup -1.1.0` are released. Magit itself will only use it for a short period of time (at least a released version of Magit, it existed since the beginning of 2014), and it can be used by other packages but Magit. Even after Magit has moved on to something else. ("magit-popup", `transient`, and `hydra` all serve similar yet different purposes). So the `magit-` prefix doesn't seem like such a good idea.

Any suggestions for a new name? It should also work as the prefix of symbols. While I don't think the term "popup" is all that great I think it is to late to lose it in this implementation, so `popup.el` would be fine - but that already exists. Maybe embrace the "popup" terminology and also acknowledge the magit heritage; and call it something like  `mpopup` or `mopup`. Or since I am not going to use that for the successor, use `idefix` (which sounds like "infix" (as in "infix arguments" and is the dog of Obelix).

**Comment by yyr** [comment]
how about `mapop.el --- magit like popup`?

**Comment by Silex** [comment]
First of all, congrats on being almost ready for releasing `next`! I'm excited about soon being able to use magit's MELPA version :)

About popup's name, I'm not a fan of idefix because it's not descriptive, if anything it suggests that it "fix IDEs" :)

I understand the idea behind `transient` as those popups are about offering choices to the user followed by actions, but again it's not very descriptive.

So, my 0.02$ would be some name that talks about actions & options. Maybe simply `action.el`? `choicable.el`? `ask-and-do.el`? :)

It's funny because we all understand clearly what this library is about, yet we all fail to find a simple explanation (and name) for it.

**Comment by tarsius** [comment]
Well I like to think of this as a synthesis of prefix arguments and prefix command:

## prefix command

```
        /- command1
prefix --- command2
        \- command3
```

Prefix commands are just like nested keymaps, except either worse or better depending what you focus on.  Better because they have a name, worse because re-evaluting `define-prefix-command` overrides customizations.

## prefix arguments

```
         /- command1
C-u ... --- command2
         \- well *any* command
```

## hydra

```
command1 -\   /- command1
command2 ---+--- command2
            ^ \- command3
            |       |
            '-------'
```

Like prefix *commands* but unlike those with an actually useful extra.

## magit-popup, transient

```
         /- argument1 -\ /- suffix1
prefix----- argument2 --+-- suffix2
       ^ \- argument3 -/
       |        |
       '--------'
```

Like prefix *arguments* but one does not have to remember for each and every command what the prefix arguments do, making it possible to offer more variants.  Also little used variants don't go completely unused because nobody can ever remember how many prefix arguments exactly are required.

On the other hand once the popup command (the prefix) has been invoked one is limited to the defined infixes and suffixes.  Then again how often does one press `C-u` without knowing yet what command one wants to pass that argument to?

So I feel "transient" hits it on the head and will probably go with `transient.el`.  But there is still time to change my mind about that.  Back to the package formerly-known-as-magit-popup.

I like `mapop`.  I like `makey` better but that's already used by @mickeynp's fork of `magit-key-mode`.  I still like `idefix` what does it matter that it doesn't *mean* anything?

> So, my 0.02$ would be some name that talks about actions & options. Maybe simply `action.el`? `choicable.el`? `ask-and-do.el`? :)

I don't see how those are more descriptive than `transient`, they just focus on something else.  `prefix-infix-and-suffix` would be more descriptive, but no.  I also want to move away from "actions" and instead speak of "commands" like everywhere else in Emacs. `ask-and-do` sound like some variation of `yes-or-no-p`.

> It's funny because we all understand clearly what this library is about, yet we all fail to find a simple explanation (and name) for it.

Yeah :-) But what about:


This library implements a generic interface for toggling switches
and setting options and then invoking an Emacs command which does
something with these arguments.  The prototypical use is for the
command to call an external process, passing on the arguments as
command line arguments.

With the Emacs concept of "prefix arguments" in mind this could be
described as "infix arguments with feedback in a buffer".  Invoking a
popup command shows its infix arguments and suffix commands in a popup
buffer.  The user then optionally changes arguments and finally
invokes one of the suffix commands which then consumes the arguments.

**Comment by Silex** [comment]
Awesome post :+1:

Whatever you decide wil lbe ok of course :-)

**Comment by tarsius** [comment]
I am gonna stick with `magit-popup`.

**Comment by tarsius** [comment]
**I am going release `2.1.0` on the 1th of July!**

Originally I planned to switch Melpa to the `next` branch about one or two weeks before the release. But now I have decided to do that only 1-3 days before the release.

The last few weeks have been very quiet and I don't expect many new issues to be reported before the release. Of course currently all users of the `next` branch made a constant decision to switch to that version, so when most other users are automatically moving the `2.1.0` some issues will come to light that only affect less enthusiastic users. 1-3 days are not enough time for all these issues to come to light but neither are 1-2 weeks. Delaying the `2.1.0` release while most users are effectively already using that release is not really a good solution to that problem anyway - and it would be rather confusing.

Instead I am prepared to release `2.1.1` or even `2.2.0` soon after `2.1.0`. But delaying the release a few days has the advantage that third-party extensions and themes can also switch to their "magit-next" branch before the actual release happens, without the maintainers having to synchronize with me and one another. And if there happen to be grave flaws in the update instructions, then a short delay also buys as some time.

----

The `magit.{org,texi}` manual currently still lives in the webpage repository. I am going to import it into the source repository sometime this weekend. At that point please start proof-reading it. If you find mistakes please don't split them into atomic commits and instead create big commits with messages such as "improve the manual". You only need to edit the org files, I will generate the texi files and amend to you commit.

Originally I intended to release it under the *GPL v2 and only v2* so that I could seed it with documentation from the git man pages. I have not done that and instead used doc-strings from magit itself. Those are released under the *GPL v3 or later*, so I am using the same license for the manual. I am the author of most of the imported doc-strings, so I could relicense them as I wish - but this is probably not the case for each and every of the important doc-strings and I didn't want to spend time investigating this. So I am using *GPL v3+*.

But I might at some point want to go back to *GPL v2-only* (to be able to imports of git's documentation). So if you don't intend to give me the permission to do so when and if I do so, then please don't contribute to the manual.

**Comment by jwiegley** [comment]
Congrats, @tarsius, and thank you so much for all of your hard work!  Magit is one of those packages in Emacs that I simply cannot do without, and that I use multiple times a day, every day.  Every minute you spend improving it is work that I make use of, and appreciate on a constant basis.

**Comment by Silex** [comment]
Hehe, @jwiegley said it all! Many, many thanks for all your hard work @tarsius and congratulations on shipping! Magit allowed me to become very fluid with operations others wouldn't dare considering for their normal workflow while feeling very safe.

I'd say that most of the operations I do *constantly* and takes about 5 seconds with magit would take from 30 seconds to 1 minute with the command line... add the time I gained every day over the last 3 years and that probably sums up to *months* of free time overall.

For example, stuffs like "reword HEAD~5 while preserving merges", "split this commit in 3 like a ninja", "stage only this line in a fixup commit, repeat 10 times, when finished `r e C-c C-c` DONE". The last one is probably the biggest time saver.

**Comment by halbtuerke** [comment]
:clap:

**Comment by andschwa** [comment]
Absolutely agree with these fellows. The ability to mark regions to stage and unstage (rather than going through `git add -p` and splitting the hunks over and over, and then going back through to make sure I didn't miss anything, and then having to `git diff --cached` to see what's staged for sure), and stashing the remainder is a lifesaver.

I **love** Magit.

**Comment by bbatsov** [comment]
This year's 1st of July will eclipse the 4th of July! :-)

I'm super excited about Magit 2.1!

**Comment by tarsius** [comment]
Thanks guys. So I figure you  don't think the [new extended introduction](http://magit.vc/about.html) is not over the top then :-)

**Comment by bbatsov** [comment]
I love it!

On 19 June 2015 at 19:06, Jonas Bernoulli <notifications@github.com> wrote:

> Thanks guys. So I figure you don't think the new extended introduction
> <http://magit.vc/about.html> is not over the top then :-)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/magit/magit/issues/1645#issuecomment-113558692>.
>



-- 
Best Regards,
Bozhidar Batsov

http://www.batsov.com

**Comment by halbtuerke** [comment]
> So I figure you don't think the new extended introduction is over the top then :-)

The best Magit introduction I have read so far!

**Comment by madscientist** [comment]
Looks great, but I see a few typos... would you like notes, and if so how do you want to get them?

**Comment by tarsius** [comment]
> Looks great, but I see a few typos... would you like notes, and if so how do you want to get them?

See the second part of https://github.com/magit/magit/issues/1645#issuecomment-112861141, i.e. wait until Sunday, then create commits that fix many issues at once.

**Comment by bradleywright** [comment]
:+

**Comment by tarsius** [comment]
I have decided to **not** make the `2.0.50` release available on Melpa before the actual release of `2.1.0`. In other words the release will happen for everyone, including Melpa users, on the 1th of July.

How this happens is quite simple. I merge `next` into `master` and then create the `2.1.0` tag. Then I push the branch and the tag. The merge causes Melpa to switch to the new version and the tag does the same for Melpa-Stable. Then I also upload to Marmalade.

**Comment by tarsius** [comment]
Oh, and I have imported all the manuals info the source repository now. That's their home now, I will occasionally copy them into the webpage repository.

**Comment by tarsius** [comment]
I have move all libraries to the new `lisp/` subdirectory. I am not a big fan of subdirectories, but here it was needed.

You'll have to adjust `load-path` accordingly. On a related note, you might also want to adjust `Info-directory-list`. See http://magit.vc/manual/magit/Installing-from-the-Git-repository.html (right now this still displays the old version, there seems to be some caching going on).

**Comment by manuel-uberti** [comment]
Just want to say thanks for the hard work. Currently using the next branch so to fix my settings accordingly. Great job on the website too.

**Comment by tarsius** [comment]
old, completed milestones

## DONE Milestone 1

* [x] Add an accurate doc-string to every command and every option.
* [x] Add a basic though incomplete manual.
* [x] Finalize key bindings.

## DONE Milestone 2

When this milestone is complete, I will feel comfortable advising people to give `next` a try. And maintainers of third-party extensions and of themes should be encouraged to take action.  Users who use such extensions and themes might want to wait until the themes and extensions have been adjusted. 

* [x] Complete the user documentation (as in everything is mentioned, it doesn't have to be perfect)
* [x] Document important apis.
  * [x] How to call git.
  * [x] How to create popups.
  * [x] How to create sections.
  * [x] Instructions for theme authors.
* [x] Completely theme the "light and dark default themes", zenburn and solarized.

## DONE Milestone 3

Many people start using `next`.  Bugs are found.  Things get fixed.

* [x] Fix all known performance issues (which can feasibly be fixed now, not #1425).
* [x] Fix all known bugs (which I can reproduce).
* [x] Improve and import manual.

**Comment by tarsius** [comment]
If someone feels like fixing mistakes in the manual - now is the time...
Thanks!

**Comment by tarsius** [comment]
I have now tested what happens when a user updates from a Melpa `1.4`-snapshot to a `2.1`-snapshot without first uninstalling `magit`, `git-commit-mode`, and `git-rebase-mode`.

If she did not use Magit before updating, then warnings about `git-commit-mode` and `git-rebase-mode` are displayed as intended. But Magit appears to be fully functional.

If she did use Magit before updating, then the installation also succeeds but no warning are displayed and trying to use `magit-status` results in

```
Debugger entered--Lisp error: (wrong-type-argument stringp t)
  call-process("git" nil (t nil) nil "--no-pager" "--literal-pathspecs" "-c" "core.preloadindex=true" "describe" "--long" "--tags" t)
  apply(call-process "git" nil (t nil) nil ("--no-pager" "--literal-pathspecs" "-c" "core.preloadindex=true" "describe" "--long" "--tags" t))
  process-file("git" nil (t nil) nil "--no-pager" "--literal-pathspecs" "-c" "core.preloadindex=true" "describe" "--long" "--tags" t)
  apply(process-file "git" nil (t nil) nil ("--no-pager" "--literal-pathspecs" "-c" "core.preloadindex=true" "describe" "--long" "--tags" t))
  (progn (apply (function process-file) magit-git-executable nil (list t nil) nil (magit-process-git-arguments args)) (if (bobp) nil (goto-char (point-min)) (buffer-substring-no-properties (point) (line-end-position))))
  (unwind-protect (progn (apply (function process-file) magit-git-executable nil (list t nil) nil (magit-process-git-arguments args)) (if (bobp) nil (goto-char (point-min)) (buffer-substring-no-properties (point) (line-end-position)))) (and (buffer-name temp-buffer) (kill-buffer temp-buffer)))
  (save-current-buffer (set-buffer temp-buffer) (unwind-protect (progn (apply (function process-file) magit-git-executable nil (list t nil) nil (magit-process-git-arguments args)) (if (bobp) nil (goto-char (point-min)) (buffer-substring-no-properties (point) (line-end-position)))) (and (buffer-name temp-buffer) (kill-buffer temp-buffer))))
  (let ((temp-buffer (generate-new-buffer " *temp*"))) (save-current-buffer (set-buffer temp-buffer) (unwind-protect (progn (apply (function process-file) magit-git-executable nil (list t nil) nil (magit-process-git-arguments args)) (if (bobp) nil (goto-char (point-min)) (buffer-substring-no-properties (point) (line-end-position)))) (and (buffer-name temp-buffer) (kill-buffer temp-buffer)))))
  magit-git-str("describe" "--long" "--tags" nil t)
  (let ((it (magit-git-str "describe" "--long" "--tags" (and (eq with-distance (quote dirty)) "--dirty") rev))) (if it (progn (let ((save-match-data-internal (match-data))) (unwind-protect (progn (string-match "\\(.+\\)-\\(?:0[0-9]*\\|\\([0-9]+\\)\\)-g[0-9a-z]+\\(-dirty\\)?$" it) (if with-distance (cons ... ...) (match-string 1 it))) (set-match-data save-match-data-internal (quote evaporate)))))))
  magit-get-current-tag(t)
  magit-insert-status-tags-line()
  run-hooks(magit-status-sections-hook)
  magit-refresh-status()
  apply(magit-refresh-status nil)
  magit-mode-refresh-buffer()
  magit-mode-init("/home/jonas/" magit-status-mode magit-refresh-status)
  apply(magit-mode-init "/home/jonas/" (magit-status-mode magit-refresh-status))
  magit-status("/home/jonas/")
  call-interactively(magit-status record nil)
  command-execute(magit-status record)
  execute-extended-command(nil "magit-status")
  call-interactively(execute-extended-command nil nil)
  command-execute(execute-extended-command)
```

After restarting Emacs, the warnings are displayed and Magit works. (It might still be a good idea to uninstall Magit, restart, and then re-install).

**Comment by thomasf** [comment]
Hooray.. been using `next` it for a day or so and it works great for me.. One difference I have found that old magit worked if I launch magit-status from dired when pwd is an symliked directory and the .git directory is not found without traversing pwd first.

**Comment by tarsius** [comment]
> One difference/minor issue I have found is that magit-status status cannot be lauched from dired if pwd is a symliked directory and the .git directory is not found without resolving pwd first.

I will have to look at this. Probably after the release. Please open a new issue.

> Also w (reword) in the commit menu does not bring up an editor instead it just finishes directly saying "git commit --amend --only" in the messages buffer. 

Does it say `0 git commit --amend --only` or `run git commit --amend --only`? In rare cases it can happen that the buffer visiting `COMMIT_EDITMSG` is created but not being displayed (in which case you have to switch to it manually.

**Comment by thomasf** [comment]
No edit buffer,.. the *magit-process buffer actually contains some possibly relevant info.

```
128 git … commit --amend --only
fatal: invalid date format: 2 hours ago
```

**Comment by tarsius** [comment]
`(magit-rev-format "%cd")` should return something like "Wed Jul 1 15:19:18 2015 +0200". If it doesn't, then you can work around the issue by setting `magit-commit-reword-override-date` to nil.

**Comment by thomasf** [comment]
Doent work either way... 

It's this entry in my global gitconfig that causes it..
```
[log]
	date = relative
``

From git show man page:
```
           ·   %cd: committer date (format respects --date= option)
           ·   %cD: committer date, RFC2822 style
```

Maybe a switch should be made to the other one in this case?

**Comment by tarsius** [comment]
Yes, `(magit-rev-format "%cd")`  would continue to respect that git variable. But setting the elisp variable should cause that to be irrelevant because `(magit-rev-format "%cd")` would not be executed anymore.

Globally set variables can be overridden for magit, by setting `magit-git-global-arguments`.

**Comment by thomasf** [comment]
`magit-commit-reword-override-date` was always nil all the time when I tested.. My magit-next is maybe two days old, dunno if that matter.  I must hurry some other things now.. I'll figure out whats really going on later.

**Comment by tarsius** [comment]
I have just pushed a wip commit with preliminary release notes to the `next` branch. It would be great if someone could fix some of the contained mistakes - am afraid there are many. Unfortunately this has to happen within the next hour, two hours at most.

@npostavs @Silex @dunn are you online and have some time you could spend on this? Thanks!

If nobody can help me with this right now, then that's fine too. I am gonna improve this during the next two hours myself too.

**Comment by npostavs** [comment]
I'll take a look

**Comment by tarsius** [comment]
@npostavs, @dunn is it okay if I squash your corrections into a single commit authored by me. 

(Eventually. For now I keep the original commits).

**Comment by dunn** [comment]
Sure.

**Comment by tarsius** [comment]
Thanks for all the corrections.

High-speed conflict resolution is exhausting!

**Comment by tarsius** [comment]
Any other incoming corrections? Else I am going to squash now.

**Comment by npostavs** [comment]
> High-speed conflict resolution is exhausting!

In retrospect, we probably could have coordinated better.

**Comment by Silex** [comment]
Congratulations on shipping! :+1: :smile:

**Comment by andschwa** [comment]
All you fans that love Magit so much and want to thank @tarsius for his *awesome* work, don't forget you can **[donate!](http://magit.vc/donations.html)**

**Comment by tarsius** [comment]
Thanks everyone!

**Comment by jcromartie** [comment]
What happened to magit-branch-manager?

**Comment by npostavs** [comment]
> What happened to magit-branch-manager?

https://github.com/magit/magit/wiki/FAQ#where-is-the-branch-manager-

**Comment by jcromartie** [comment]
Oh, thanks. I was searching for "magit-branch-manager" specifically :)

**Comment by tarsius** [comment]
I have just added a big pointing finger to the readme, the wiki, and the homepage. I hope it helps.
